File: src/main/java/me/nasukhov/TukitaLearner/db/StringListConverter.java
package me.nasukhov.TukitaLearner.db;

import jakarta.persistence.AttributeConverter;
import jakarta.persistence.Converter;
import java.util.Arrays;
import java.util.List;

@Converter
public class StringListConverter implements AttributeConverter<List<String>, String> {
    private static final String SEPARATOR = ";";

    @Override
    public String convertToDatabaseColumn(List<String> list) {
        if (list == null || list.isEmpty()) {
            return null;
        }

        return String.join(SEPARATOR, list);
    }

    @Override
    public List<String> convertToEntityAttribute(String dbData) {
        if (dbData == null || dbData.isEmpty()){
            return List.of();
        }

        return Arrays.asList(dbData.split(SEPARATOR));
    }
}

File: src/main/kotlin/me/nasukhov/TukitaLearner/db/StringListConverter.kt
package me.nasukhov.TukitaLearner.db

import jakarta.persistence.AttributeConverter
import jakarta.persistence.Converter

@Converter
class StringListConverter : AttributeConverter<List<String>, String> {
    companion object {
        private const val SEPARATOR = ";"
    }

    override fun convertToDatabaseColumn(list: List<String>?): String? {
        if (list == null || list.isEmpty()) {
            return null
        }

        return list.joinToString(SEPARATOR)
    }

    override fun convertToEntityAttribute(dbData: String?): MutableList<String> {
        if (dbData == null || dbData.isEmpty()) {
            return mutableListOf()
        }

        return dbData.split(SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }.toMutableList()
    }
}
